src/SFA.DAS.ApprenticeCommitments/Application/Commands/SendInvitationReminders/SendInvitationRemindersCommandHandler.cs

```diff
@@ -0,0 +1,76 @@
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using MediatR;
+using Microsoft.Extensions.Logging;
+using SFA.DAS.ApprenticeCommitments.Apis.InnerApi;
+using SFA.DAS.ApprenticeCommitments.Application.Services;
+using SFA.DAS.ApprenticeCommitments.Application.Services.ApprenticeLogin;
+
+namespace SFA.DAS.ApprenticeCommitments.Application.Commands.SendInvitationReminders
+{
+    public class SendInvitationRemindersCommandHandler : IRequestHandler<SendInvitationRemindersCommand>
+    {
+        private readonly ApprenticeCommitmentsService _apprenticeCommitmentsService;
+        private readonly ApprenticeLoginService _apprenticeLoginService;
+        private readonly CommitmentsV2Service _commitmentsV2Service;
+        private readonly ILogger<SendInvitationRemindersCommandHandler> _logger;
+
+        public SendInvitationRemindersCommandHandler(
+            ApprenticeCommitmentsService apprenticeCommitmentsService,
+            ApprenticeLoginService apprenticeLoginService, 
+            CommitmentsV2Service commitmentsV2Service,
+            ILogger<SendInvitationRemindersCommandHandler> logger)
+        {
+            _apprenticeCommitmentsService = apprenticeCommitmentsService;
+            _apprenticeLoginService = apprenticeLoginService;
+            _commitmentsV2Service = commitmentsV2Service;
+            _logger = logger;
+        }
+
+        public async Task<Unit> Handle(SendInvitationRemindersCommand command, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Getting Registrations which are due a reminder");
+            var list = (await _apprenticeCommitmentsService.GetReminderRegistrations()).Registrations;
+
+            // This just wouldn't work 
+            //Parallel.ForEach(list, new ParallelOptions {MaxDegreeOfParallelism = 4}, async registration =>
+            //{
+            //    await SendReminder(registration);
+            //});
+
+            await Task.WhenAll(list.Select(x=>SendReminder(x, command.SendNow)));
```

Comment:
Is there a danger of bombarding the inner APIs and login service with a spike of HTTP requests if there are many apprenticeships that need reminding in a given day?  

Given a large enough number of reminders it sounds like the kind of thing where the first so many could succeed but then the latter part of the list fail.

Reply:
This is also my concern. I worked a long time trying to user Parallel.ForEach() and throttling it but the call completed before the async tasks finished. It's worthwhile throttling this